---
title: "Competitive_mapping_microbiome"
author: "Bosco Gracia Alvira"
date: "2024-08-29"
output: html_document
---

```{r setup}
knitr::opts_knit$set(root.dir = "/Users/bgracia/Dropbox (PopGen)/Bosco/PhD_Dropbox/Ancestral_microbiome/data/Competitive_mapping_microbiome")
knitr::opts_chunk$set(echo = FALSE)
library(lattice)
library(data.table)
library(Biostrings)
library(ggplot2)
library(dplyr)
library(tidyverse)
library(stringr)
library(gtools)
library(pheatmap)

```


```{r Loading the data}

# Paths to the files that we want to import
reads_path <- "/Users/bgracia/Dropbox (PopGen)/Bosco/PhD_Dropbox/Ancestral_microbiome/data/Competitive_mapping_microbiome/reads_mapped.tsv"
uniq_path <- "/Users/bgracia/Dropbox (PopGen)/Bosco/PhD_Dropbox/Ancestral_microbiome/data/Competitive_mapping_microbiome/uniq_mapped.tsv"
size_path <- "/Users/bgracia/Dropbox (PopGen)/Bosco/PhD_Dropbox/Ancestral_microbiome/data/Competitive_mapping_microbiome/genome_size.tsv"
visuals_path <- "/Users/bgracia/Dropbox (PopGen)/Bosco/PhD_Dropbox/Ancestral_microbiome/visuals/"

# Load the coverage, sort the samples alphanumerically and make them row names
coverage <- fread(uniq_path, header=T)
coverage$Sample <- factor(coverage$Sample , levels = mixedsort(coverage$Sample))
coverage <- as.data.frame(coverage[order(coverage$Sample), ])
row.names(coverage) <- coverage$Sample
coverage$Sample <- NULL
coverage <- as.data.frame(t(coverage))

# Load genome sizes info, turn genome names into row names, turn the sizes into numbers and transform them to kb
size <- as.data.frame(fread(size_path))
size$Size <- gsub(",", "", size$Size)
size$Size <- as.numeric(size$Size)
size_vec <- as.vector(size[, 2])

```


```{r Reads normalisation to tpm}

# We normalise the data using tpm

# First we divide the number of reads mapped to each genome by the genome size, and multiply by 1000 (kb)
size_norm <- sweep(coverage, 1, size_vec, FUN = "/") * 1e3

# Counts normalised by genome size are then multiplied by 10^6 and divided by the total number of reads in each sample
tpm <- as.data.frame(t(t(size_norm)*1e6/ colSums(size_norm)))

# As a result, all columns (samples) add up to 10^6, regardless of their coverage
colSums(tpm)

```


```{r Data subsets}

# We divide the table into isolates and pools columns

# Subset columns that start with "i"
tpm_i <- tpm %>% select(starts_with("i"))
tpm_i <- rownames_to_column(tpm_i, var = "Taxon")

# Subset columns that do not start with "i"
tpm_pools <- tpm %>% select(-starts_with("i"))
tpm_pools <- rownames_to_column(tpm_pools, var = "Taxon")

# Melt the dataframe to long format
tpm_i_long <- reshape2::melt(tpm_i, id.vars = "Taxon", variable.name = "Sample", value.name = "TPM")
tpm_pools_long <- reshape2::melt(tpm_pools, id.vars = "Taxon", variable.name = "Sample", value.name = "TPM")

# Based on the sample name, we create two new columns: Generation and Temperature
tpm_i_long <- tpm_i_long %>%
  mutate(
    generation = as.numeric(str_extract(Sample, "(?<=F)\\d+")),
    temperature = case_when(
      str_detect(Sample, "h") ~ "hot",
      str_detect(Sample, "c") ~ "cold",
      str_detect(Sample, "i") ~ "isolate")
  )

tpm_pools_long <- tpm_pools_long %>%
  mutate(
    generation = as.numeric(str_extract(Sample, "(?<=F)\\d+")),
    temperature = case_when(
      str_detect(Sample, "h") ~ "hot",
      str_detect(Sample, "c") ~ "cold",
      str_detect(Sample, "i") ~ "isolate")
  )

# Plot the distribution of counts
ggplot(tpm_i_long, aes(x = TPM)) +
  geom_histogram(fill = "blue", color = "black", alpha = 0.7) +
  theme_minimal() +
  labs(title = "Distribution of TPM Counts", x = "TPM", y = "Frequency")

# How many elements have less than 10000 reads?
tpm_i_long %>% filter(TPM < 10000) %>% nrow()
nrow(tpm_i_long)

# Plot the distribution of counts
ggplot(tpm_pools_long, aes(x = TPM)) +
  geom_histogram(fill = "blue", color = "black", alpha = 0.7) +
  theme_minimal() +
  labs(title = "Distribution of TPM Counts", x = "TPM", y = "Frequency")

# How many elements have less than 10000 TPMs in the pools?
tpm_pools_long %>% filter(TPM < 10000) %>% nrow()
nrow(tpm_pools_long)

```

```{r Merge low abundance taxa}

# How many taxa have less than 10000 TPMs in ALL the samples (less than 1% abundance)?
# We extract the names of these taxa
sum(apply(tpm_pools[,-1], 1, function(row) all(row < 10000)))
other_taxa <- apply(tpm_pools[,-1], 1, function(row) all(row < 10000))
other_taxa <- tpm_pools[other_taxa, 1]

tpm_pools_long$Taxon <- ifelse(tpm_pools_long$Taxon %in% other_taxa, "Other", tpm_pools_long$Taxon)
tpm_pools_long <- aggregate(TPM ~ Sample + Taxon + generation + temperature, data = tpm_pools_long, sum)

```

```{r Taxons order and colour, message=FALSE, warning=FALSE}
# Here we set the order in which we want the taxons to appear.
Amplicon_level <- c("Other",
                    "Corynebacteriales",
                    "Chloroplast",
                    "Lactobacillales",
                    "Chitinophagales",
                    "Orbales",
                    "Enterobacterales",
                    "ASV_7",
                    "ASV_5",
                    "ASV_4",
                    "ASV_2",
                    "ASV_15")

taxon_levels <- c("Other",
                  "o_Borreliales",
                  "s_Corynebacterium_provencense",
                  "s_Corynebacterium_nuruki",
                  "g_Enteroccus",
                  "s_Enterococcus_faecalis",
                  "s_Leuconostoc_pseudomesenteroides",
                  "s_Levilactobacillus_brevis",
                  "s_Lactiplantibacillus_paracasei",
                  "s_Lactiplantibacillus_plantarum",
                  "g_Chitinophaga",
                  "f_Enterobacteriaceae_3",
                  "f_Enterobacteriaceae_2",
                  "f_Enterobacteriaceae_1",
                  "f_Enterobacteriaceae",
                  "s_Morganella_morganii",
                  "s_Acetobacter_indonesiensis_3",
                  "s_Acetobacter_indonesiensis_2",
                  "s_Acetobacter_indonesiensis_1",
                  "s_Acetobacter_sicerae",
                  "s_Acetobacter_oryzifermentans",
                  "g_Acetobacter")

# Here we choose the colour 
Amplicon_colours <- c("#E69F00", #12
                      "#0072B2", #11
                      "#D55E00", #10
                      "#FF3030", #9
                      "#CC79A7", #8
                      "#56B4E9", #7
                      "#F0E442", #6
                      "#458B00", #5
                      "#00CDCD", #4
                      "#68228B", #3
                      "#FF1493", #2
                      "#FFD700") #1


Abundance_colours <- c("#E69F00", #12
                       "#0072B2", #11
                       "#0072B2", #11
                       "#FF3030", #10
                       "#FF3030", #10
                       "#FF3030", #10
                       "#FF3030", #10
                       "#FF3030", #10
                       "#FF3030", #10
                       "#CC79A7", #8
                       "#56B4E9", #7
                       "#56B4E9", #7
                       "#F0E442", #6
                       "#F0E442", #6
                       "#458B00", #5
                       "#458B00", #5
                       "#458B00", #5
                       "#00CDCD", #4
                       "#68228B", #3 
                       "#68228B") #3


# This palette clusters the taxa by colour

Amplicon_colours <- c("#E69F00", #12
                      "#FFF0F5", #11
                      "#D55E00", #10
                      "#FF3030", #9
                      "#CC79A7", #8
                      "#6B8E23", #7
                      "#B3EE3A", #6
                      "#00CDCD", #5
                      "#00F5FF", #4
                      "#B2DFEE", #3
                      "#27408B", #2
                      "#63B8FF") #1


Abundance_colours <- c("#E69F00", #12
                       "#FFF0F5", #11
                       "#FFF0F5", #11
                       "#FF3030", #10
                       "#FF3030", #10
                       "#FF3030", #10
                       "#FF3030", #10
                       "#FF3030", #10
                       "#FF3030", #10
                       "#CC79A7", #8
                       "#6B8E23", #7
                       "#6B8E23", #7
                       "#B3EE3A", #6
                       "#B3EE3A", #6
                       "#00CDCD", #5
                       "#00CDCD", #5
                       "#00CDCD", #5
                       "#00F5FF", #4
                       "#B2DFEE", #3  
                       "#B2DFEE") #3
```







```{r Trajectory over the time}

# The table has information from 2 temperature time series. We create a list in which each series is an element.
tpm_list <- split(tpm_pools_long, tpm_pools_long$temperature)
tpm_list <- tpm_list[names(tpm_list) != "isolateNA"]
plot_list <- list()

# We iterate to create a ggplot object for each series
for(i in seq_along(tpm_list)) {
  # Create a plot for each data frame
 p <- ggplot(tpm_list[[i]], aes(x = generation, y = TPM, fill = Taxon)) +
    geom_area() +
    ylab("Taxon relative Abundance") +
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    xlab("Generation") +
    labs(fill = "Taxon", title = paste("Sample", names(tpm_list)[i]))
  
   p
   #ggsave(filename = paste0(visuals_path,"Lpla_3strains_bbmap_",names(df_list)[i],".png"), width = 6, height = 4, dpi = 300)
  
  # Store the plot in the list
  plot_list[[i]] <- p
}

for(p in plot_list) {
  print(p)
}

```
