---
title: "Competitive_mapping_microbiome"
author: "Bosco Gracia Alvira"
date: "2024-08-29"
output: html_document
---

```{r setup}
knitr::opts_knit$set(root.dir = "/Users/bgracia/Dropbox (PopGen)/Bosco/PhD_Dropbox/Ancestral_microbiome/data/Lpla_dynamics")
knitr::opts_chunk$set(echo = FALSE)
library(lattice)
library(data.table)
library(Biostrings)
library(ggplot2)
library(dplyr)
library(tidyverse)
library(stringr)
library(gtools)
library(pheatmap)

```


```{r Loading the data}

# Paths to the files that we want to import
reads_path <- "/Users/bgracia/Dropbox (PopGen)/Bosco/PhD_Dropbox/Ancestral_microbiome/data/Lpla_dynamics_replicates/reads_mapped.tsv"
uniq_path <- "/Users/bgracia/Dropbox (PopGen)/Bosco/PhD_Dropbox/Ancestral_microbiome/data/Lpla_dynamics_replicates/uniq_mapped.tsv"
size_path <- "/Users/bgracia/Dropbox (PopGen)/Bosco/PhD_Dropbox/Ancestral_microbiome/data/Lpla_dynamics_replicates/genome_size.tsv"
metadata_isolates_path <- "/Users/bgracia/Dropbox (PopGen)/Bosco/PhD_Dropbox/Ancestral_microbiome/data/Lpla_dynamics_replicates/metadata.tsv"
visuals_path <- "/Users/bgracia/Dropbox (PopGen)/Bosco/PhD_Dropbox/Ancestral_microbiome/visuals/Lpla_dynamics_replicates"

# Load the coverage, sort the samples alphanumerically and make them row names
coverage <- fread(uniq_path, header=T)
coverage$Sample <- factor(coverage$Sample , levels = mixedsort(coverage$Sample))
coverage <- as.data.frame(coverage[order(coverage$Sample), ])
row.names(coverage) <- coverage$Sample
coverage$Sample <- NULL
coverage <- as.data.frame(t(coverage))

# Load genome sizes info, turn genome names into row names, turn the sizes into numbers and transform them to kb
size <- as.data.frame(fread(size_path))
size$Size <- gsub(",", "", size$Size)
size$Size <- as.numeric(size$Size)
size_vec <- as.vector(size[, 2])

# Create a colour palette for our temperatures
temp_palette <- c("HOT" = "red", "BASE" = "grey", "COLD" = "lightblue")

```


```{r Reads normalisation to tpm}

# We normalise the data using tpm

# First we divide the number of reads mapped to each genome by the genome size, and multiply by 1000 (kb)
size_norm <- sweep(coverage, 1, size_vec, FUN = "/") * 1e3

# Counts normalised by genome size are then multiplied by 10^6 and divided by the total number of reads in each sample
tpm <- as.data.frame(t(t(size_norm)*1e6/ colSums(size_norm)))

# As a result, all columns (samples) add up to 10^6, regardless of their coverage
colSums(tpm)

```


```{r Data subsets}

# We divide the table into isolates and pools columns

# Subset columns that start with "i"
tpm_i <- tpm %>% select(starts_with("i"))
tpm_i <- rownames_to_column(tpm_i, var = "Strain")

# Subset columns that do not start with "i"
tpm_pools <- tpm %>% select(-starts_with("i"))
tpm_pools <- rownames_to_column(tpm_pools, var = "Strain")

# Melt the dataframe to long format
tpm_i_long <- reshape2::melt(tpm_i, id.vars = "Strain", variable.name = "Sample", value.name = "TPM")
tpm_pools_long <- reshape2::melt(tpm_pools, id.vars = "Strain", variable.name = "Sample", value.name = "TPM")

# Set the factor levels of the strains
tpm_i_long$Strain <- factor(tpm_i_long$Strain, levels = c("S103","S239","B89"))
tpm_pools_long$Strain <- factor(tpm_pools_long$Strain, levels = c("S103","S239","B89"))

# Based on the sample name, we create two new columns: Generation and Temperature
tpm_pools_long <- tpm_pools_long %>%
  mutate(
    generation = as.numeric(str_extract(Sample, "(?<=F)\\d+")),
    replicate = str_extract(Sample, "(?<=_)\\w+$"),
    temperature = case_when(
      str_detect(Sample, "h") ~ "hot",
      str_detect(Sample, "c") ~ "cold",
      str_detect(Sample, "i") ~ "isolate")
  )

# Plot the distribution of counts
ggplot(tpm_i_long, aes(x = TPM)) +
  geom_histogram(fill = "blue", color = "black", alpha = 0.7) +
  theme_minimal() +
  labs(title = "Distribution of TPM Counts", x = "TPM", y = "Frequency")

# How many elements have less than 10000 reads?
tpm_i_long %>% filter(TPM < 10000) %>% nrow()
nrow(tpm_i_long)

# Plot the distribution of counts
ggplot(tpm_pools_long, aes(x = TPM)) +
  geom_histogram(fill = "blue", color = "black", alpha = 0.7) +
  theme_minimal() +
  labs(title = "Distribution of TPM Counts", x = "TPM", y = "Frequency")

# How many elements have less than 10000 TPMs in the pools?
tpm_pools_long %>% filter(TPM < 10000) %>% nrow()
nrow(tpm_pools_long)

```

```{r Trajectory over the time-replicates (pools)}

# The table has information from 20 different time series (2 regimes * 10 replicates).
#We create a list in which each series is an element.

# Create a unique identifier for each time series (TemperatureReplicate)
tpm_pools_long$list <- paste0(tpm_pools_long$temperature,tpm_pools_long$replicate)

# Split the database based on this identifier
tpm_list <- split(tpm_pools_long, tpm_pools_long$list)
tpm_list <- tpm_list[names(tpm_list) != "isolateNA"]
plot_list <- list()

# We iterate to create a ggplot object for each series
for(i in seq_along(tpm_list)) {
  
  # Open a graphics device (e.g., PNG)
  png(filename = paste0("Pools_",names(tpm_list)[i],".png"))
  
  
  # Create a plot for each data frame
  p <- ggplot(tpm_list[[i]], aes(x = generation, y = TPM, fill = Strain)) +
      geom_area() +
      ylab("Strain relative Abundance") +
      scale_fill_manual(values = c("S239" = "red", "S103" = "lightblue", "B89" = "grey")) +
      theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
      xlab("Generation") +
      labs(fill = "Strain", title = paste("Sample", names(tpm_list)[i]))

  ggsave(filename = paste0("Pools_",names(tpm_list)[i],".png"), plot = p, path = visuals_path, width = 12, height = 6, dpi = 300)
  
  # Close the graphics device
  dev.off()

  # Store the plot in the list
  plot_list[[i]] <- p
}

for(p in plot_list) {
  print(p)
}

```


```{r Trajectory over the time - temperature}

# In this chunk we merge the samples from all the replicates.

tpm_pools_long_temp <- tpm_pools_long %>%
  group_by(Strain, generation, temperature) %>%
  summarize(
    mean_TPM = mean(TPM, na.rm = TRUE),
    sd_TPM = sd(TPM, na.rm = TRUE),
    .groups = 'drop'
  )

# The table has information from 2 different time series, one per regime.
#We create a list in which each series is an element.

# Split the database based on the temperature regime
tpm_list <- split(tpm_pools_long_temp, tpm_pools_long_temp$temperature)

tpm_list <- tpm_list[names(tpm_list) != "isolateNA"]
plot_list <- list()

# We iterate to create a ggplot object for each series
for(i in seq_along(tpm_list)) {
  
  # Open a graphics device (e.g., PNG)
  png(filename = paste0("Pools_",names(tpm_list)[i],".png"))
  
  
  # Create a plot for each data frame
  p <- ggplot(tpm_list[[i]], aes(x = generation, y = mean_TPM, fill = Strain)) +
      geom_area() +
      ylab("Strain relative Abundance") +
      scale_fill_manual(values = c("S239" = "red", "S103" = "lightblue", "B89" = "grey")) +
      theme(axis.text.x = element_text(angle = 90, hjust = 1, size=15)) +
      xlab("Generation") +
      labs(fill = "Strain", title = paste("Sample", names(tpm_list)[i]))

  ggsave(filename = paste0("Pools_",names(tpm_list)[i],".png"), plot = p, path = visuals_path, width = 12, height = 6, dpi = 300)
  
  # Close the graphics device
  dev.off()

  # Store the plot in the list
  plot_list[[i]] <- p
}

for(p in plot_list) {
  print(p)
}

```


```{r Isolates' strain composition}

# Import the metadata information of the isolates
metadata_isolates <- fread(metadata_isolates_path, header=T)

# Rename sample to Sample to match the tpm Sample names
metadata_isolates <- metadata_isolates %>%
                        select(sample, Temperature, Generation) %>%
                        dplyr::rename(Sample = sample) %>%
                        mutate(Sample = paste0("i",Sample)) %>%
                        mutate(Source = "Isolate")

# Merge TPM and metadata
tpm_i_long_merged <- merge(tpm_i_long,metadata_isolates, by="Sample")

# Open a graphics device (e.g., PNG)
png(filename = "Isolates_composition.png")

# Plot the strain abundance of each isolate
Isolates_plot <- ggplot(tpm_i_long_merged, aes(x = Sample, y = TPM, fill = Strain)) +
                      geom_bar(stat = "identity", position = "stack") +
                      ylab("Strain relative Abundance") +
                      scale_fill_manual(values = c("S239" = "red", "S103" = "lightblue", "B89" = "grey")) +
                      theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
                      xlab("Generation") +
                      labs(fill = "Strain", title = "Sample")

  ggsave(filename = "Isolates_composition.png", plot = Isolates_plot, path = visuals_path, width = 15, height = 6, dpi = 300)
  
  # Close the graphics device
  dev.off()

Isolates_plot
```

