---
title: "SNPs_plotting"
author: "Bosco Gracia Alvira"
date: "2024-07-25"
output: html_document
---

```{r Choose the taxon of interest}
knitr::opts_chunk$set(echo = FALSE)

# Get the command line arguments
args <- commandArgs(trailingOnly = TRUE)

# The first argument will be the value of $REPLY
reply <- args[1]
#reply <- "s__Lactiplantibacillus_plantarum"

# Print the value of reply (for debugging purposes)
print(paste("The following taxon will be used:", reply))

knitr::opts_knit$set(root.dir = paste0("/Users/bgracia/PopGen Dropbox/Martin McFly/Bosco/PhD_Dropbox/Ancestral_microbiome/data/",reply,"/SNPs_analysis"))

```


```{r Loading packages}
library(lattice)
library(data.table)
library(ggplot2)
library(dplyr)
library(tidyverse)
library(stringr)
library(VariantAnnotation)
library(qvalue)

```

```{r Loading the data}

# Paths to the files that we want to import
data_path <- paste0("/Users/bgracia/PopGen Dropbox/Martin McFly/Bosco/PhD_Dropbox/Ancestral_microbiome/data/SNPs_analysis/",reply)
metadata_isolates_path <- "/Users/bgracia/PopGen Dropbox/Martin McFly/Bosco/PhD_Dropbox/Ancestral_microbiome/data/SNPs_analysis/metadata.tsv"
visuals_path <- "/Users/bgracia/PopGen Dropbox/Martin McFly/Bosco/PhD_Dropbox/Ancestral_microbiome/visuals/SNPs_analysis"

# Read the vcf files with and without annotations
vcf <- readVcf(paste0(data_path,"/",reply,".vcf"))
annotations <- readVcf(paste0(data_path,"/",reply,"_annotated.vcf"))

# Read the allele frequency data from each sample
AD_df <- as.data.frame(geno(vcf)$AD)
AD_df$rowname <- row.names(AD_df)
AD_df <- AD_df %>%
          separate(rowname, into = c("CHROM","POS_REF_ALT"), sep = ":") %>%
          separate(POS_REF_ALT, into = c("POS","REF_ALT"), sep = "_", extra = "merge") %>%
          separate(REF_ALT, into = c("REF", "ALT"), sep = "/") %>%
          dplyr::select(CHROM, POS, REF, ALT, everything())
          

# Remove four isolates that I know that are contaminated and have intermediate allele frequencies
if (reply == "s__Lactiplantibacillus_plantarum") {
  AD_df <- AD_df %>%
    dplyr::select(-iX79, -iX81, -iX595, -iX371)
}

```


```{r Process the frequency table}

# Transform the data into a data.table and melt the dataframe to long format
AD_df <- as.data.table(AD_df)
freq_long <- data.table::melt(AD_df, id.vars = colnames(AD_df)[1:4], variable.name = "Sample", value.name = "AF")

# Split the AF column into reference (RO) and alternative (AF)
freq_long <- freq_long %>%
  mutate(RF = map_int(AF, 1),
         AF = map_int(AF, 2))

# Calculate the summed allele depth (SAD) and relative frequencies from absolute number of counts
freq_long <- freq_long[, "SAD" := (RF + AF)]
freq_long <- freq_long[, `:=`(RF = RF / (RF + AF), AF = AF / (RF + AF))]

# Create a new column with unique position information
freq_long <- freq_long %>%
  mutate(CHROM_POS = paste0(CHROM,":",POS))

# Make the table wide again
freq_wide <- freq_long %>%
  dplyr::select(Sample, CHROM_POS, RF) %>%  # Select relevant columns
  pivot_wider(names_from = CHROM_POS, values_from = RF) %>%  # Pivot to wide format
  column_to_rownames(var = "Sample")  # Set Sample as row names

# Remove the columns (SNPs) that are NaN in at least 1 sample
rows_freq <- row.names(freq_wide)
freq_wide <- as.data.table(freq_wide)
freq_wide <- freq_wide[, lapply(.SD, function(x) if (any(is.nan(x))) NULL else x), .SDcols = names(freq_wide)]
freq_wide <- as.data.frame(freq_wide)
row.names(freq_wide) <- rows_freq
snps <- as.numeric(ncol(freq_wide))

# Transform data to arcosin
freq_asin <- 2*asin(sqrt(freq_wide))

```

```{r SFS calculation}

# We calculate the site frequency spectrum for each sample, as well as the median allele frequency
SFS <- freq_long[,c("Sample","RF")] %>%
          mutate(RF = ifelse(RF > 0.5, 1 - RF, RF)) %>%
          na.omit()

SFS <- SFS %>%
  group_by(Sample) %>%
  mutate(medRF = median(as.numeric(RF))) %>%
  mutate(allele_count = n()) %>%
  mutate(Title = gsub("_contigs$", "", Sample)) %>%
  ungroup()

median_SFS <- SFS %>%
  group_by(Sample) %>%
  summarize(
    median_SFS = median(as.numeric(RF), na.rm = TRUE),
    Total = sum(as.numeric(RF)),
    No0 = sum(as.numeric(RF) != 0, na.rm = TRUE),
    percent = No0/Total
  ) %>%
  dplyr::rename(name = Sample)

ggplot(median_SFS, aes(x = name, y = percent)) +
  geom_bar(stat = "identity") +
  labs(title = "Number of Non-Zero Values per Sample",
       x = "Sample Name",
       y = "Number of variant sites") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))

# Open a graphics device (e.g., PNG)
png(filename = paste0("SFS_",reply,".png"))

# Plot the SFS for each isolate
SFS_plot <- ggplot(SFS, aes(x = RF)) +
                geom_histogram(binwidth = 0.1, boundary = 0, fill = "blue", color = "black") +
                facet_wrap(~ Title, scales = "free_y") +
                geom_vline(aes(xintercept = medRF), color = "red", linetype = "dashed") +
                labs(title = paste0(reply,", SFS (Folded)"), x = "Allele Frequency (0-0.5)", y = "Count") +
                theme_minimal() +
                theme(
                  #axis.title.x = element_blank(),
                  #axis.title.y = element_blank(),
                  axis.text.x = element_blank(),
                  axis.text.y = element_blank(),
                  #axis.ticks.x = element_blank(),
                  #axis.ticks.y = element_blank()
                ) +
                facet_wrap(~ Title, scales = "free_y", labeller = labeller(Title = function(x) {
                  paste0(x," (",SFS %>% filter(Title == x) %>% pull(allele_count) %>% unique(),")")
                }))

# Save the plot using ggsave
ggsave(filename = paste0("SFS_",reply,".png"), plot = SFS_plot, path = visuals_path, width = 8, height = 6, dpi = 300)

# Close the graphics device
dev.off()

# Plot the SFS also in R
SFS_plot

```


```{r Import metadata}

# Import the metadata of all the isolates and subset those present in the PCA
metadata_isolates <- fread(metadata_isolates_path, header=T)
metadata_isolates <- metadata_isolates %>%
                        dplyr::select(sample, Temperature, Generation) %>%
                        dplyr::rename(name = sample) %>%
                        mutate(name = paste0("i",name))
metadata_isolates$Source <- "Isolate"

metadata_pools <- as.data.frame(rownames(freq_asin)[grepl("^cF|^hF", rownames(freq_asin))])
colnames(metadata_pools) <- "name"
metadata_pools <- metadata_pools %>%
  mutate(
    Temperature = case_when(
      str_detect(name, "F0") ~ "BASE",
      str_detect(name, "^h") ~ "HOT",
      str_detect(name, "^c") ~ "COLD"),
    Generation = str_extract(name, "F\\d+"),
    Source = "Pool")

metadata <- rbind(metadata_isolates,metadata_pools)
metadata <- merge(metadata, median_SFS, by = "name")

```


```{r Calculate PCA}

# Do the PCA 
pca.res <- prcomp(freq_asin, retx=TRUE, center = FALSE, scale. =FALSE)

# The variable pca.res$x contains the coordinates of each sample in each PC
pca_data <- as.data.frame(pca.res$x)

# Add sample information (assuming row names of freq_asin are sample names)
pca_data$name <- rownames(freq_asin)

# How much variance is explained by each PC? 
# The variance explained by each PC is the square of the standard deviations divided by the total variance.
sdev <- pca.res$sdev
variance_explained <- sdev^2 / sum(sdev^2)
percentage_variance_explained <- variance_explained * 100

```


```{r Plot PCAs}

# Merge the eigenvectors and the metadata
pca_merged <- merge(pca_data, metadata, by = "name")

# Remove "_contigs$" from the end of some genomes
pca_merged$name <- gsub("_contigs$", "", pca_merged$name)

pca_pool <- subset(pca_merged, pca_merged$Source=="Pool")
pca_isolate <- subset(pca_merged, pca_merged$Source=="Isolate")

# Open a graphics device (e.g., PNG)
png(filename = paste0("PCA_temp_",reply,".png"))

# Plot the PCA colouring the samples by isolation temperature regime
PCA_temp <- ggplot() +
                geom_point(data = pca_isolate, aes(x = PC1, y = PC2, color = Temperature, shape = Source), size = 2, alpha = 0.9) +
                geom_text(data = pca_isolate, aes(x = PC1, y = PC2, label = name, color = Temperature), size = 1.5, hjust = -0.5) +
                geom_point(data = pca_pool, aes(x = PC1, y = PC2, color = Temperature, shape = Source), size = 2) +
                geom_text(data = pca_pool, aes(x = PC1, y = PC2, label = name, color = Temperature), size = 1.5, hjust = -0.5) +
                labs(
                  title = paste0(reply," (",snps," variable SNPs)"),
                  x = paste0("PC1 (", round(percentage_variance_explained[1], 2), "% variance)"),
                  y = paste0("PC2 (", round(percentage_variance_explained[2], 2), "% variance)"),
                  color = "Temperature", shape = "Source") +
                scale_colour_manual(values = c("HOT" = "red", "BASE" = "grey", "COLD" = "lightblue", "COLD-CONSTANT" = "blue")) +
                theme_minimal()

# Save the plot using ggsave
ggsave(filename = paste0("PCA_temp_",reply,".png"), plot = PCA_temp, path = visuals_path, width = 6, height = 4, dpi = 300)

# Close the graphics device
dev.off()

# Open a graphics device (e.g., PNG)
png(filename = paste0("PCA_SFS_",reply,".png"))

# Plot the PCA colouring the samples by median AF
PCA_SFS <- ggplot() +
              geom_point(data = pca_isolate, aes(x = PC1, y = PC2, color = median_SFS, shape = Source), size = 1, alpha = 0.4) +
              geom_text(data = pca_isolate, aes(x = PC1, y = PC2, label = name, color = median_SFS), size = 1.5, hjust = -0.5) +
              geom_point(data = pca_pool, aes(x = PC1, y = PC2, color = median_SFS, shape = Source), size = 2) +
              geom_text(data = pca_pool, aes(x = PC1, y = PC2, label = name, color = median_SFS), size = 1.5, hjust = -0.5) +
              labs(
                title = paste0(reply," (",snps," variable SNPs)"),
                x = paste0("PC1 (", round(percentage_variance_explained[1], 2), "% variance)"),
                y = paste0("PC2 (", round(percentage_variance_explained[2], 2), "% variance)"),
                color = "median_SFS", shape = "Source") +
              scale_color_gradient(low = "purple", high = "yellow") +
              theme_minimal()

# Save the plot using ggsave
ggsave(filename = paste0("PCA_SFS_",reply,".png"), plot = PCA_SFS, path = visuals_path, width = 6, height = 4, dpi = 300)

# Close the graphics device
dev.off()

# Plot here the PCAs as well
PCA_temp
PCA_SFS

```

```{r Annotation of SNPs}

# Read the annotations for each variant and select the columns of interest
annotation_df <- as.data.frame(info(annotations))
annotation_df <- annotation_df %>%
                      dplyr::select(IsSynonymous,IsTransition,IsGenic,AminoAcidChange,Gene,Product) %>%
                      rownames_to_column(var = "CHROM_POS") %>%
                      mutate(ntChange=str_sub(CHROM_POS, -3, -1)) %>% 
                      mutate(CHROM_POS=str_sub(CHROM_POS, 1, -5)) %>% 
                      mutate(across(where(is.list), ~ map_chr(., as.character)))
            
annotated_variants <- merge(freq_long, annotation_df, by="CHROM_POS")
meta_snps <- merge(annotated_variants, metadata, by.x="Sample", by.y="name")

```


```{r Identify genotype-specific SNPs}

if (reply == "s__Lactiplantibacillus_plantarum") {

# Subset the isolates, create the Strain variable with the genotype of each isolate
meta_i <- meta_snps %>%
                filter(Source == "Isolate") %>%
                filter(RF != "NaN") %>%
                mutate(Genotype = case_when(
                        Sample %in% c("iB25", "iX76", "iX371") ~ "S239",
                        Temperature == "HOT" ~ "S239",
                        Temperature == "BASE" ~ "B89",
                        Temperature == "COLD" ~ "S103",
                        Temperature == "COLD-CONSTANT" ~ "S103")) 

# Proportion of the genomes from each strain that has the SNP
p <- meta_i %>%
  group_by(CHROM_POS, Genotype, Gene, Product) %>%
  summarize(p = mean(RF, na.rm = TRUE)) %>%
  pivot_wider(names_from = Genotype, values_from = p, names_prefix = "p_")

# Number of genomes per strain
N <- meta_i %>%
  group_by(CHROM_POS, Genotype, Gene, Product) %>%
  summarize(N = n()) %>%
  pivot_wider(names_from = Genotype, values_from = N, names_prefix = "N_")

# Expected proportion of genomes that should contain the SNP under a uniform distribution (null expectation)
E <- meta_i %>%
  mutate(Presence = if_else(RF > 0.5, 1, 0)) %>%
  group_by(CHROM_POS) %>%
  summarize(E = mean(Presence))

# Join the three tables and calculate the associated_groups to each SNP, that don't follow the null expectation
# Some SNPs are not covered in all the genomes (they are indels in some), so these I remove them as well
pN <- p %>%
  left_join(N, by = c("CHROM_POS", "Product")) %>%
  rowwise() %>%
  left_join(E, by = "CHROM_POS") %>% 
  mutate(associated_groups = paste(names(dplyr::select(cur_data(),
        starts_with("p_")))[which(c_across(starts_with("p_")) > E)],
        collapse = ",")) %>%
  ungroup() %>%
  dplyr::select(-E) %>%
  dplyr::mutate(accession = CHROM_POS) %>% 
  dplyr::select(CHROM_POS, accession, Product, associated_groups, p_B89, N_B89, p_S239, N_S239, p_S103, N_S103) %>% 
  drop_na()

# Save the file
write_tsv(pN, "~/pN.tsv")

enrichment_script <- "/Users/bgracia/PopGen Dropbox/Martin McFly/Bosco/PhD_Dropbox/Ancestral_microbiome/code_new/anvi-script-enrichment-stats"

args <- c(paste0("--input=~/pN.tsv"), "--output=~/enriched_SNPs.tsv")
system2(enrichment_script, args)

enriched_snps <- fread("~/enriched_SNPs.tsv", header=T)
enriched_snps <- enriched_snps %>% 
                    mutate(adjusted_q_value = as.numeric(adjusted_q_value)) %>% 
                    filter(adjusted_q_value <= 0.05) %>%
                    dplyr::select(-accession,-Product) %>% 
                    left_join(annotation_df, by = "CHROM_POS")

# Save the file
write_tsv(enriched_snps, paste0(data_path,"/enriched_snps.tsv"))

# Now we have our enriched table. We can remove the input and output from the home directory
system2("rm", c("~/pN.tsv", "~/enriched_SNPs.tsv"))


# For each SNP we want to turn the private allele into the reference one
# To do this we have to modify several fields of the table for the SNPs that are associated with two groups:

# This function swaps the last character with the last third character (reference and alternative alleles CHROM_POS)
swap_nt <- function(x) {
  ref <- str_sub(x, 1, 1)
  alt <- str_sub(x, 3, 3)
  str_sub(x, 3, 3) <- ref
  str_sub(x, 1, 1) <- alt
  x
}

swap_aa <- function(x) {
  ref <- str_sub(x, 1, 1)
  alt <- str_sub(x, -1, -1)
  str_sub(x, 1, 1) <- alt
  str_sub(x, -1, -1) <- ref
  x
}

str_sub(enriched_snps$ntChange[18385], 3, 3)
str_sub(enriched_snps$AminoAcidChange[18385], -1, -1)

# 
transform_associated_groups <- function(groups) {
  case_when(
    groups == "p_B89,p_S103" ~ "p_S239",
    groups == "p_B89,p_S239" ~ "p_S103",
    groups == "p_S103,p_S239" ~ "p_B89",
    groups == "p_B89,p_S103,p_S239" ~ "ALL",
    TRUE ~ groups
  )
}

# Modify the columns as specified
enriched_snps_2 <- enriched_snps %>%
  mutate(
    ntChange = if_else(str_detect(associated_groups, ","), swap_nt(ntChange), ntChange),
    AminoAcidChange = if_else(str_detect(associated_groups, ","), swap_aa(AminoAcidChange), AminoAcidChange),
    p_B89 = if_else(str_detect(associated_groups, ","), abs(p_B89 - 1), p_B89),
    p_S239 = if_else(str_detect(associated_groups, ","), abs(p_S239 - 1), p_S239),
    p_S103 = if_else(str_detect(associated_groups, ","), abs(p_S103 - 1), p_S103),
    inversed = if_else(str_detect(associated_groups, ","), "inv", NA),
    associated_groups = transform_associated_groups(associated_groups),
    Colour = case_when(
                          associated_groups=="p_B89" ~ "green",
                          associated_groups=="p_S239" ~ "red",
                          associated_groups=="p_S103"~ "blue"),
  )


enriched_snps %>%
  summarise(na_count = sum(is.na(Colour)))

enriched_snps %>%
  filter((associated_groups == "p_B89" | associated_groups == "p_S103,p_S239"), IsGenic==1) %>%
  summarise(count = n())

enriched_snps %>%
  filter((associated_groups == "p_S103" | associated_groups == "p_B89,p_S239"), IsGenic==1) %>%
  summarise(count = n())

enriched_snps %>%
  filter((associated_groups == "p_S239" | associated_groups == "p_B89,p_S103"), IsSynonymous==0) %>%
  summarise(count = n())

meta_pool <- meta_snps %>%
                filter(Source == "Pool") %>%
                filter(IsSynonymous == "0") %>%
                filter(Temperature != "COLD") %>%
                mutate(Generation = as.numeric(sub("^F", "", Generation))) %>% 
                left_join(enriched_snps_2, by = "CHROM_POS") %>%
                mutate (AF = if_else(inversed == "inv", abs(AF - 1), AF),
                        RF = if_else(inversed == "inv", abs(RF - 1), RF))


ggplot(meta_pool, aes(x = Generation, y = RF, group = CHROM_POS, colour=Colour)) +
              geom_line(linewidth=0.05) +
              scale_colour_identity()

}

```






```{r SNPs trajectory}
# We want to plot the SNPs trajectories in the hot and the cold regime
meta_pool <- meta_snps %>%
                filter(Source == "Pool") %>%
                filter(IsSynonymous == "0") %>%
                filter(Temperature != "COLD") %>%
                mutate(Generation = as.numeric(sub("^F", "", Generation)))

ggplot(meta_pool, aes(x = Generation, y = RF, group = CHROM_POS)) +
              geom_line(linewidth=0.01)


```

